File: leetcode/121_Best_Time_to_Buy_and_Sell_Stock.go
package main

/*You are given an array prices where prices[i] is the price of a given stock on the ith day.

You want to maximize your profit by choosing a single day to buy one stock and choosing a different day in the future to sell that stock.

Return the maximum profit you can achieve from this transaction. If you cannot achieve any profit, return 0.*/

func maxProfit(prices []int) int {
	minPrice := prices[0]
	maxProfit := 0

	for _, price := range prices {
		if price < minPrice {
			minPrice = price
		} else if price-minPrice > maxProfit {
			maxProfit = price - minPrice
		}
	}
	return maxProfit
}

func main() {
	// prices := []int{7, 1, 5, 3, 6, 4}

	// prices := []int{7, 6, 4, 3, 1}

	prices := []int{2, 4, 1}
	println(maxProfit(prices))
}

File: leetcode/121_Best_Time_to_Buy_and_Sell_Stock.rs

struct Solution {}

impl Solution {
    pub fn max_profit(prices: Vec<i32>) -> i32 {
        let mut min_price = prices[0];
        let mut max_profits = 0;

        for i in 1..prices.len() {
            if prices[i] < min_price {
                min_price = prices[i];
            } else if prices[i] - min_price > max_profits {
                max_profits = prices[i] - min_price;
            }
        }

        return max_profits;
    }
}

fn main() {
    let prices = vec![7, 1, 5, 3, 6, 4];
    let max_profits = Solution::max_profit(prices);
    println!("{}", max_profits);
}
